src/neo/Wallets/Wallet.cs

```diff
@@ -527,22 +527,33 @@ public long CalculateNetworkFee(DataCache snapshot, Transaction tx)
             int size = Transaction.HeaderSize + tx.Signers.GetVarSize() + tx.Attributes.GetVarSize() + tx.Script.GetVarSize() + IO.Helper.GetVarSize(hashes.Length);
             uint exec_fee_factor = NativeContract.Policy.GetExecFeeFactor(snapshot);
             long networkFee = 0;
+            int index = -1;
             foreach (UInt160 hash in hashes)
             {
+                index++;
                 byte[] witness_script = GetAccount(hash)?.Contract?.Script;
+                byte[] invocationScript = null;
 
-                if (witness_script is null && tx.Witnesses != null)
+                if (tx.Witnesses != null)
                 {
-                    // Try to find the script in the witnesses
-
-                    foreach (var witness in tx.Witnesses)
+                    if (witness_script is null)
                     {
-                        if (witness.ScriptHash == hash)
+                        // Try to find the script in the witnesses
+
+                        foreach (var witness in tx.Witnesses)
```

Comment:
Why do we have this loop here? Normal transactions have 1:1 mapping between senders and witnesses, but this implies that this rule can be broken. Maybe I'm missing something in client-wallet interactions in C# code, but I'd expect client to respect this 1:1 mapping even if it doesn't provide full invocation scripts in some cases (transaction is not signed yet).

Reply:
Done